src/Microsoft.ML.Core/Data/IEstimator.cs

```diff
@@ -30,18 +30,18 @@ public enum VectorKind
 
             public readonly string Name;
             public readonly VectorKind Kind;
-            public readonly DataKind ItemKind;
+            public readonly ColumnType ItemType;
```

Comment:
>ColumnType ItemType [](start = 28, length = 19)

This is somewhat problematic for key columns. How do we represent the two distinct concepts of a key-columns of known and unknown counts?

Reply:
Well, `ItemType` is never supposed to be a key. For example, for a scalar key column, the correct representation is `Kind = Scalar, ItemType = PrimitiveType.U4, IsKey = true`.
I don't think we should even have key types of unknown counts: it is already causing issues for `Term` transform, and I'm yet to see any benefit of this type.

---
In reply to: [213151407](https://github.com/dotnet/machinelearning/pull/753#discussion_r213151407) [](ancestors = 213151407)

Reply:
I agree that keys of unknown size should go away, but are we going to do this in this PR?

That we're representing a key-type using a column type which is not a key-type at all, and whose only connection to it is that their `DataKind` happen to be the same, is rather odd and unfortunate. I hope we can imagine a better way here, though I'm not sure I see one right off the bat.

---
In reply to: [213156549](https://github.com/dotnet/machinelearning/pull/753#discussion_r213156549) [](ancestors = 213156549,213151407)

Reply:
We are also representing a vector type using a column type which is not a vector-type at all, and whose only connection to it is that their `ItemType` happen to be the same :)

---
In reply to: [214125274](https://github.com/dotnet/machinelearning/pull/753#discussion_r214125274) [](ancestors = 214125274,213156549,213151407)